src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.cpp

```diff
@@ -151,7 +151,64 @@ bool ShootOrPassPlayFSM::passFound(const Update& event)
 
 bool ShootOrPassPlayFSM::shouldAbortPass(const Update& event)
 {
-    // TODO (#2384): implement this
+    // lost possession
+    if (event.common.world.getTeamWithPossession() != TeamSide::FRIENDLY)
+    {
+        return true;
+    }
+
+    // calculate an inflated polygon along the vector from the passer to receiver
+    const auto ball_position  = event.common.world.ball().position();
+    const auto passer_point   = best_pass_and_score_so_far.pass.passerPoint();
+    const auto receiver_point = best_pass_and_score_so_far.pass.receiverPoint();
+
+    const double RADIUS = 0.5;
+
+    Vector place_to_ball = passer_point.toVector() - receiver_point.toVector();
+    Vector ball_to_place = -place_to_ball;
+
+    Point place_l = passer_point + (place_to_ball.normalize(RADIUS) +
+                                    place_to_ball.perpendicular().normalize(RADIUS));
+    Point place_r = passer_point + (place_to_ball.normalize(RADIUS) -
+                                    place_to_ball.perpendicular().normalize(RADIUS));
+
+    Point ball_l = receiver_point + (ball_to_place.normalize(RADIUS) +
+                                     ball_to_place.perpendicular().normalize(RADIUS));
+    Point ball_r = receiver_point + (ball_to_place.normalize(RADIUS) -
+                                     ball_to_place.perpendicular().normalize(RADIUS));
+
+    const auto pass_area_polygon = Polygon({
+                                               place_l,
+                                               place_r,
+                                               ball_l,
+                                               ball_r,
+                                           })
+                                       .expand(1);
+
+    // calculate a polygon that contains the receiver and passer point, and checks if the
+    // ball is inside it. if the ball isn't being passed to the receiver then we should
+    // abort
+    if (!contains(pass_area_polygon, ball_position))
+    {
+        return true;
+    }
+
+
+    // distance between robot and ball is too far,
+    // i.e. team might still have possession, but kicker/passer doesn't have control over
+    // ball
+    if (event.common.world.ball().velocity().length() <
+        this->ai_config.shoot_or_pass_play_config().ball_shot_threshold())
+    {
+        if ((event.common.world.ball().position() -
+             best_pass_and_score_so_far.pass.passerPoint())
+                .length() >
+            this->ai_config.shoot_or_pass_play_config().min_distance_to_pass())
+        {
+            return true;
+        }
+    }
```

Comment:
I'd prefer a variable like 
```suggestion
    bool ball_far_from_robot = (event.common.world.ball().velocity().length() <
        this->ai_config.shoot_or_pass_play_config().ball_shot_threshold()) && 
    ((event.common.world.ball().position() -
             best_pass_and_score_so_far.pass.passerPoint())
                .length() >
            this->ai_config.shoot_or_pass_play_config().min_distance_to_pass());
```
and then returning that bool